runtime/lua/vim/lsp/codelens.lua

```diff
@@ -0,0 +1,134 @@
+})
+local active_requests = {}
+local M = {}
+local namespaces = setmetatable({}, {
+  __index = function(t, key)
+    local value = api.nvim_create_namespace('vim_lsp_codelens:' .. key)
+    rawset(t, key, value)
+    return value
+  end;
+})
+
+
+--- Return all lenses for the given buffer
+-- TODO: add client_id filter?
+function M.get(bufnr)
+  local lenses_by_client = lens_cache_by_buf[bufnr]
+  if not lenses_by_client then return {} end
+  local lenses = {}
+  for _, client_lenses in pairs(lenses_by_client) do
+    vim.list_extend(lenses, client_lenses)
+  end
+  return lenses
+end
+
+
+-- todo: expose get_all() ?
+
+--- Run the code lens in the current line
+function M.run()
+  local line = api.nvim_win_get_cursor(0)[1]
+  local options = vim.tbl_filter(
+    function(lens) return lens.range.start.line == (line - 1) end,
+    M.get(api.nvim_get_current_buf())
+  )
+  if #options == 0 then
+    print('No executable codelens found at current line')
+  elseif #options == 1 then
+    vim.lsp.buf.execute_command(options[1].command)
+  else
+    local options_strings = {"Code lenses:"}
+    for i, option in ipairs(options) do
+       table.insert(options_strings, string.format('%d. %s', i, option.command.title))
+    end
+    local choice = vim.fn.inputlist(options_strings)
+    if choice < 1 or choice > #options then
+      return
+    end
+    vim.lsp.buf.execute_command(options[choice].command)
+  end
+end
+
+
+local function resolve_lens(cache, idx, lens)
+  -- If the chandler should be overridable in vim.lsp.handlers the structure of the cache would need to change
+  -- So that the closure is not necessary
+  -- Maybe {bufnr -> client_id -> linenr -> lenses}?
+  local _, cancel = vim.lsp.buf_request(0, 'codeLens/resolve', lens, function(err, _, result, client_id, bufnr)
+    assert(not err, vim.inspect(err))
+    if not result then return end
+    if result.command then
+      local line = result.range.start.line
+      local chunks = { {result.command.title, 'LspCodeLens'} }
+      local ns = namespaces[client_id]
+      api.nvim_buf_set_virtual_text(bufnr, ns, line, chunks, {})
+    end
+    cache[idx] = result
+  end)
+  table.insert(active_requests, cancel)
+end
+
+
+function M.on_codelens(err, _, result, client_id, bufnr)
+  assert(not err, vim.inspect(err))
+  -- todo, window could have changed, retrieve window using bufnr?
+  local wininfo = vim.fn.getwininfo(api.nvim_get_current_win())[1]
+  local lenses = {}
+  local ns = namespaces[client_id]
+  -- Can we avoid flickering? should we track which lens disappeared somehow?
+  api.nvim_buf_clear_namespace(bufnr, ns, 1, -1)
+  for idx, lens in pairs(result or {}) do
+    lenses[idx] = lens
+    if lens.command then
+      local line = lens.range.start.line
+      local chunks = { {lens.command.title, 'LspCodeLens'} }
+      api.nvim_buf_set_virtual_text(bufnr, ns, line, chunks, {})
+    else
+      -- todo: Should there be a different logic to decide whether to resolve the codelens?
```

Comment:
I've implemented codelenses in my vimfiles and there I resolve just before executing (would be on M.run here). It makes execution slower, but afaik I don't use any language server that supports codeLens/resolve, so not sure which approach is better or how to measure it :(

Reply:
I did that in a first version as well, but it doesn't really work for unresolved lenses.

If the lens isn't resolved already there isn't anything sensible that could be displayed to the user.
For example, language servers like eclipse.jdt.ls provide the option to enable reference counts via code lens.


The response just contains lens like this:


```
data = { "file:///path/to/file.java", {
        character = 13,
        line = 145
      }, "references" },
    range = {
      end = {
        character = 20,
        line = 145
      },
      start = {
        character = 13,
        line = 145
      }
```


The `data` part is arbitrary.

Once resolved it receives the `command` part in addition:

```
{
  command = {
    arguments = { "file:///path/to/file.java", {
        character = 16,
        line = 557
      }, { {
          range = {
            end = {
              character = 64,
              line = 55
            },
            start = {
              character = 23,
              line = 55
            }
          },
          uri = "file:///path/to/file.java"
        } } },
    command = "java.show.references",
    title = "1 reference"
  }
}
```

![image](https://user-images.githubusercontent.com/38700/101379742-e6a87300-38b4-11eb-8b3c-a14cd9f5bb7c.png)